src/attribute.rs

```diff
@@ -0,0 +1,119 @@
+
+pub trait Attribute {
+    fn parse(raw_attribute: Vec<u8>) -> Result<Self, String>
+    where
+        Self: Sized;
+}
+
+impl Attribute for String {
+    fn parse(raw_attribute: Vec<u8>) -> Result<Self, String> {
+        String::from_utf8(raw_attribute).map_err(|err| {
+            format!(
+                "parse: failed to parse selector String value, error: {}",
+                err
+            )
+        })
+    }
+}
+
+impl Attribute for i64 {
+    fn parse(raw_attribute: Vec<u8>) -> Result<Self, String> {
+        if raw_attribute.len() != 8 {
+            return Err(format!(
+                "Int value expected to be 8 bytes, but got {}",
+                raw_attribute.len()
+            ));
+        }
+        Ok(i64::from_le_bytes(
+            raw_attribute[..8]
+                .try_into()
+                .expect("This has to be 8 bytes long!"),
+        ))
+    }
+}
+
+impl Attribute for u64 {
+    fn parse(raw_attribute: Vec<u8>) -> Result<Self, String> {
+        if raw_attribute.len() != 8 {
+            return Err(format!(
+                "UInt value expected to be 8 bytes, but got {}",
+                raw_attribute.len()
+            ));
+        }
+        Ok(u64::from_le_bytes(
+            raw_attribute[..8]
+                .try_into()
+                .expect("This has to be 8 bytes long!"),
+        ))
+    }
+}
+
+impl Attribute for f64 {
+    fn parse(raw_attribute: Vec<u8>) -> Result<Self, String> {
+        if raw_attribute.len() != 8 {
+            return Err(format!(
+                "Float value expected to be 8 bytes, but got {}",
+                raw_attribute.len()
+            ));
+        }
+        Ok(f64::from_le_bytes(
+            raw_attribute[..8]
+                .try_into()
+                .expect("This has to be 8 bytes long!"),
+        ))
+    }
+}
+
+impl Attribute for Vec<u8> {
+    fn parse(raw_attribute: Vec<u8>) -> Result<Self, String> {
+        Ok(raw_attribute)
+    }
+}
+
+impl Attribute for bool {
+    fn parse(raw_attribute: Vec<u8>) -> Result<Self, String> {
+        if raw_attribute.len() != 1 {
+            return Err(format!(
+                "Bool value expected to be 1 byte, but got {}",
+                raw_attribute.len()
+            ));
+        }
+        Ok(raw_attribute[0] & 1 == 1)
+    }
+}
+
+impl Attribute for DateTime<FixedOffset> {
+    fn parse(raw_attribute: Vec<u8>) -> Result<Self, String> {
+        if raw_attribute.len() != 8 {
+            return Err(format!(
+                "parse: Timestamp expected to be 8 bytes, but got {}",
+                raw_attribute.len()
+            ));
+        }
+
+        let nanos = i64::from_le_bytes(
+            raw_attribute.as_slice()[..8]
+                .try_into()
+                .expect("This has to be 8 bytes long!"),
+        );
+        Ok(DateTime::from_timestamp_nanos(nanos).into())
+    }
+}
+
+#[allow(dead_code)]
+pub fn get_attribute<T>(f: &Filter, attr: &str) -> Result<T, String>
```

Comment:
This is apparently still `dead_code`, but I also wonder how this is going to be useful... `<T>` needs to be known at compile time here, so that we _need_ to know at method invocation time what the `type` of `attr`... is this ever going to be the case that this is known _before_ runtime?

Reply:
As an example when I was building some response stuff for auth I was using it like this (providing a default to infer types):

```
attr.set_source(build_peer(
    get_attribute(&self, "source.address").unwrap_or("".to_string()),
    get_attribute(&self, "source.port").unwrap_or(0i64) as u32,
));
```

But then again now I think about it, these should probably remain unset if we can't get the attribute and not be a default value.

Reply:
Yeah, that actually makes sense... forgot that in the case of auth, there is a set of attributes (both their name _and_ type) that need to be retrieved no matter what... so yes, that `fn` would be useful for that. The `unwrap_or` I'm less confident about it indeed as well